overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,384 @@
+	serialReq.WaitFor(genKey)
+	retrieveSerial := m.state.NewTask("retrieve-n-install-serial", i18n.G("Retrieve and install device serial"))
+	retrieveSerial.WaitFor(serialReq)
+	serialReq.Set("serial-setup-task", retrieveSerial.ID())
+
+	chg := m.state.NewChange("become-operational", i18n.G("Setting up device identity"))
+	chg.AddAll(state.NewTaskSet(genKey, serialReq, retrieveSerial))
+
+	return nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *DeviceManager) Ensure() error {
+	err := m.ensureOperational()
+	if err != nil {
+		return err
+	}
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait implements StateManager.Wait.
+func (m *DeviceManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop implements StateManager.Stop.
+func (m *DeviceManager) Stop() {
+	m.runner.Stop()
+}
+
+var (
+	keyLength = 4096
+	// XXX: a 2nd different URL for nonce?
+	// TODO: this will come as config from the gadget snap
+	serialRequestURL = "https://serial.request" // XXX dummy value!
+)
+
+func doGenerateDeviceKey(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+	// XXX: where to store device private key ultimately? use asserts support?
+	var encoded string
+	err := st.Get("device-keypair", &encoded)
+	if err == nil {
+		// nothing to do
+		return nil
+	}
+	if err != state.ErrNoState {
+		return err
+	}
+	keyPair, err := rsa.GenerateKey(rand.Reader, keyLength)
+	if err != nil {
+		return fmt.Errorf("cannot generate device key pair: %v", err)
+	}
+
+	encoded = base64.StdEncoding.EncodeToString(x509.MarshalPKCS1PrivateKey(keyPair))
+	st.Set("device-keypair", encoded)
+	return nil
+}
+
+func keyPair(st *state.State) (*rsa.PrivateKey, error) {
+	var encoded string
+	err := st.Get("device-keypair", &encoded)
+	if err == state.ErrNoState {
+		return nil, fmt.Errorf("internal error: cannot find device key pair")
+	}
+	if err != nil {
+		return nil, err
+	}
+	b, err := base64.StdEncoding.DecodeString(encoded)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: cannot decode device key pair: %v", err)
+	}
+	pk, err := x509.ParsePKCS1PrivateKey(b)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: cannot decode device key pair: %v", err)
+	}
+	return pk, nil
+}
+
+type serialSetup struct {
+	Serial        string `json:"serial"`
+	SerialRequest string `json:"serial-request"`
+}
+
+type nonceTicketResp struct {
+	NonceTicket string `json:"nonce-ticket"`
+}
+
+var errPoll = errors.New("serial-request accepted, poll later")
+
+func submitSerialRequest(cli *http.Client, encodedSerialRequest []byte) (*asserts.Serial, error) {
+	resp, err := cli.Post(serialRequestURL, asserts.MediaType, bytes.NewBuffer(encodedSerialRequest))
+	if err != nil {
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
+		return nil, fmt.Errorf("unexpected status code %d trying to post serial request to %s)", resp.StatusCode, serialRequestURL)
```

Comment:
How are these cases any different?  This could easily be a hiccup in the server, or a proxy error, or something else. Feels like we have a lot of logic trying to pick particular cases, but we only really have two situations: we have a serial response, or we don't. If we don't, we need to retry.